src/properties/array_property.rs

```diff
@@ -75,6 +75,7 @@ impl ArrayProperty {
         cursor: &mut R,
         hints: &HashMap<String, String>,
         properties_stack: &mut Vec<String>,
+        large_world_coordinates: bool,
```

Comment:
Rather than passing it to the property, the R should be an instance of GvasTrait + GvasReader and the property should query large_world_coordinates from that instead.

Reply:
Coupling a `Read` instance with data extracted from the underlying file may not be the best approach. This would complicate the API for `*Property::read` and `*Property::write` functions, and cause inconsistencies with their `GvasFile` and `GvasHeader` equivalents. With that said, I do think the argument list is getting to be quite verbose, and one way to clean that up would be to use a struct to wrap the options:

```rust
struct PropertyOptions {
    hints: &HashMap<String, String>,
    properties_stack: &mut Vec<String>,
    large_world_coordinates: bool,
}
```

That way, a single `&PropertyOptions` reference can be passed to all `::read` and `::write` functions, limiting the surface area that would be impacted by any future changes.

Reply:
I don't think `large_world_coordinates` belong in property options, it's more a gvas file setting, that is not necessarily related to how a property is read, but may affect it, it would also be beneficial to store it all in one place, rather than having it in both options and the gvas file itself.

Reply:
Regarding it complicating API for `*Property::read/write`, I don't think it will complicate it, GvasReader would just pass the cursor functions through, therefore the function usage and code inside of it would remain the same, just the trait bounds would be different.

Reply:
Including the LWC bool in the reader would not follow the single responsibility principle. I think that the reader should focus on the data stream, rather than the de-serialized state.

Although LWC might appear unrelated to the other struct members, all three are actually the injected dependencies for property serialization and de-serialization. Unfortunately, they are inconsistently used throughout the codebase, which results in several distinct read function prototypes and several distinct write function prototypes. I believe that having a single read function prototype and a single write function prototype would simplify the code by making it more homogenous.

Reply:
The reader is considered more of an archive because it contains the header and other stuff. Is LWC a global asset switch or can it be enabled on one property and disabled on another?

Reply:
UE5 can be configured at compile time with LWC enabled or disabled.

The value can not change from one property to the next within a given file.

Reply:
In this case it sounds like it should be on the archive rather than passed to each property

Reply:
I've shared my perspective on why the reader can't determine the header value during property reading. If you still prefer that approach, it would be helpful if you could provide some example code to illustrate its functionality, as we may not be fully understanding each other's points.

Reply:
The solution I'm proposing would look something like this:
```rust
pub trait GvasTrait {
    // Custom version getters could be here
    
    /// Current archive cursor position
    fn position(&mut self) -> u64;
    /// Seek
    fn seek(&mut self, style: SeekFrom) -> io::Result<u64>;

    // Not sure if gvas can be read with usmap, but if it can, then hints will be moved to .usmap mappings in the future
    // and this will be used as a getter
    /// Get .usmap mappings
    fn get_mappings(&self) -> Option<&Usmap>;

    fn is_lwc_enabled(&self) -> bool;
}

pub trait GvasReader : GvasTrait {
        /// Read `u8`
        fn read_u8(&mut self) -> io::Result<u8>;
        /// Read `i8`
        fn read_i8(&mut self) -> io::Result<i8>;
        /// Read `u16`
        fn read_u16<T: ByteOrder>(&mut self) -> io::Result<u16>;
        /// Read `i16`
        fn read_i16<T: ByteOrder>(&mut self) -> io::Result<i16>;
        /// Read `u32`
        fn read_u32<T: ByteOrder>(&mut self) -> io::Result<u32>;
        /// Read `i32`
        fn read_i32<T: ByteOrder>(&mut self) -> io::Result<i32>;
        /// Read `u64`
        fn read_u64<T: ByteOrder>(&mut self) -> io::Result<u64>;
        /// Read `i64`
        fn read_i64<T: ByteOrder>(&mut self) -> io::Result<i64>;
        /// Read `f32`
        fn read_f32<T: ByteOrder>(&mut self) -> io::Result<f32>;
        /// Read `f64`
        fn read_f64<T: ByteOrder>(&mut self) -> io::Result<f64>;
        /// Read a String
        fn read_string(&mut self) -> Result<String, Error>;
        /// Read an exact amount of bytes into a slice
        fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()>;
        /// Read `bool`
        fn read_bool(&mut self) -> io::Result<bool>;
}
```

Reply:
Property creation example would look something like this:
```rust
impl Int8Property {
    #[inline]
    pub(crate) fn read<R: GvasReader>(
        cursor: &mut R,
        include_header: bool,
    ) -> Result<Self, Error> {
        if include_header {
            check_size!(cursor, 1);
            let separator = cursor.read_u8()?;
            assert_eq!(separator, 0);
        }
        Ok(Int8Property {
            value: cursor.read_i8()?,
        })
    }
}
```

Reply:
I believe that implementing this trait may reveal that the LWC data's lifecycle differs from the reader's, making it unsuitable for GvasTrait to hold LWC state. Discussing hypothetical code might not be productive. I've uploaded a new commit with a v3 test file. If possible, please share a working commit that showcases your proposed architecture.

Reply:
The implementation I am suggesting has a flaw where it's not possible to do this
```rust
        let buf = &mut Cursor::new(Vec::new());
        self.write_body(buf)?;
        let buf = buf.get_ref();
```

But I am not sure how this would be fixed if mappings were to be added either, does seek really impact performance that bad?

Reply:
After further implementation, it looks like my idea works for more complex projects like `unreal_asset`, but adds unnecessary complexity to this one, so having large_world_coordinates inside property options for gvas is fine.